azurerm/resource_arm_automation_schedule.go

```diff
@@ -0,0 +1,317 @@
+package azurerm
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/arm/automation"
+	"github.com/Azure/go-autorest/autorest/date"
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/helper/validation"
+)
+
+func resourceArmAutomationSchedule() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmAutomationScheduleCreateUpdate,
+		Read:   resourceArmAutomationScheduleRead,
+		Update: resourceArmAutomationScheduleCreateUpdate,
+		Delete: resourceArmAutomationScheduleDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"account_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"resource_group_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"start_time": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"expiry_time": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"interval": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"frequency": {
+				Type:             schema.TypeString,
+				Required:         true,
+				DiffSuppressFunc: ignoreCaseDiffSuppressFunc,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(automation.Day),
+					string(automation.Hour),
+					string(automation.Month),
+					string(automation.OneTime),
+					string(automation.Week),
+				}, true),
+			},
+			"first_run": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"second": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  -1,
```

Comment:
My problem is that i cannot get the inner resource as `schema.ResourceData`. I get the `first_time` object like this: `d.Get("first_run").([]interface{})` and the items in this list are type `map[string]interface{}`. So i cannot use the Get method, i can only index in this map.

Can you help me out how can i access the `first_time` list element as `ResourceData`?

Reply:
Sure thing - from memory you should be able to access the fields via:

```
items := d.Get("first_run").([]interface{})
item := items[0].(map[string]interface{})
if v, ok := item["foo"]; ok {
  value := v.(string)
  # ...
}
```

However - thinking about this some more, I think we'd be better to remove the `first_run` block altogether in favour of making `start_time` field required? We could then push all of these fields to the top level in a data source, pseudocode below:

```
data "azurerm_scheduled_time" "test" {
  frequency = "hourly"
  hour = 5
}
```

and access via:

```
output "next_start_time" {
  value = "${data. azurerm_scheduled_time.test.next_scheduled_time}"
}

# note: this field may not necessarily make sense, just a thought
output "last_run_time" {
  value = "${data. azurerm_scheduled_time.test.last_scheduled_time}"
}
```

What do you think? If we took this route I think this'd be easier to do as a separate PR if possible, because I have a feeling we'd use this Data Source in multiple places (such as for the Azure Monitoring Rules)

Reply:
Thanks for your help.

Yep, this sounds reasonable. I almost finished with all the suggested modifications. So can we first complete this PR and if is merged then i can extract the first_time out as a Data Source and make start_time required.

What do you think is this doable plan?

Reply:
Thanks for working through the other comments - I'll take an updated look through shortly.

We'd be better to remove `first_time` and make `start_time` required before merging - at which point removing these fields becomes more complex since some folks may opt to use the resource once it's merged into master if possible?

Reply:
Yeah sure, i get your point. I will remove it shortly.

Reply:
`items := d.Get("first_run").([]interface{})
item := items[0].(map[string]interface{})
if v, ok := item["foo"]; ok {
  value := v.(string)
  # ...
}`
This solution is not really working for me. The scheduling test is failed if i use this expression. In the test there is no day_of_month property but this code snippet returns an `ok = true` with a value of `0`. That's why i use the default value -1. Do you have any other idea how to workaround this?

I know, i know ... this code will be removed from this PR but first i want to clean everything so i can save the `first_run` code for later use. :)

Reply:
That's a shame - I was hoping that `ok` would be false given it's not specified

```
if v := item["foo"].(int); v >= 0 {
  # do something with v
}
```

is probably the best we can do at this point.

When this field is in the Data Source and at the top-level in the schema `v, ok := d.GetOk("foo")` should work and ignore the default value - however when it's nested like that this isn't the case unfortunately

Reply:
Yeah right, now i pushed the code without the first_run field. I removed it hopefully from everywhere (even from docs). And i change the start_time field from optional to required. 
I alost finished with everything. I would like to check the `ExpectNonEmptyPlan: true,` cases on monday, maybe i can remove them as you suggested.  
Please do a fresh review on the code as i modify a lot of files. 
Cheers!